modules/launcher/src/main/java/org/mule/runtime/module/launcher/AbstractArtifactClassLoaderBuilder.java

```diff
@@ -0,0 +1,258 @@
+import org.mule.runtime.module.launcher.application.ArtifactPlugin;
+import org.mule.runtime.module.launcher.application.ArtifactPluginFactory;
+import org.mule.runtime.module.launcher.application.CompositeArtifactClassLoader;
+import org.mule.runtime.module.launcher.application.FilePackageDiscoverer;
+import org.mule.runtime.module.launcher.application.PackageDiscoverer;
+import org.mule.runtime.module.launcher.plugin.ArtifactPluginDescriptor;
+import org.mule.runtime.module.launcher.plugin.ArtifactPluginDescriptorLoader;
+import org.mule.runtime.module.launcher.plugin.ArtifactPluginRepository;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for all artifacts class loader filters.
+ *
+ * @param <T> the type of the filer.
+ * @since 4.0
+ */
+public abstract class AbstractArtifactClassLoaderBuilder<T extends AbstractArtifactClassLoaderBuilder>
+{
+
+    private static final String SHARED_LIB_ARTIFACT_NAME = "sharedLibs";
+
+    private final DeployableArtifactClassLoaderFactory artifactClassLoaderFactory;
+    private final PackageDiscoverer packageDiscoverer = new FilePackageDiscoverer();
+    private final ArtifactPluginRepository artifactPluginRepository;
+    private final ArtifactPluginFactory artifactPluginFactory;
+    private File[] applicationPluginDependencies = new File[0];
+    private Set<ArtifactPluginDescriptor> artifactPluginDescriptors = new HashSet<>();
+    private ArtifactPluginDescriptorLoader artifactPluginDescriptorLoader;
+    private File pluginsSharedLibFolder;
+    private String artifactId = UUID.getUUID();
+    private ArtifactDescriptor artifactDescriptor;
+    private ArtifactClassLoader parentClassLoader;
+    private Map<String, ArtifactClassLoader> artifactPluginClassLoaders = new HashMap<>();
+
+    /**
+     * Creates an AbstractArtifactClassLoaderBuilder.
+     *
+     * @param artifactClassLoaderFactory factory for the classloader specific to the artifact resource and classes
+     * @param artifactPluginRepository repository of plugins contained by the runtime
+     * @param artifactPluginFactory factory for creating artifact plugins
+     * @param artifactPluginDescriptorLoader loader for plugin zip files into their descriptors
+     */
+    public AbstractArtifactClassLoaderBuilder(DeployableArtifactClassLoaderFactory artifactClassLoaderFactory,
+                                              ArtifactPluginRepository artifactPluginRepository,
+                                              ArtifactPluginFactory artifactPluginFactory,
+                                              ArtifactPluginDescriptorLoader artifactPluginDescriptorLoader)
+    {
+        this.artifactClassLoaderFactory = artifactClassLoaderFactory;
```

Comment:
Would be good to add checks to ensure that the parameters have the expected values